### ISSUE TITLE
output_padding argument is ignored in deconvolution, leading to cudnn errors

### ISSUE THREAD

**Issue by ngimel** [pytorch/pytorch#964]
The following snippet produces an error in deconvolution:
```.py
import torch
import torch.nn as nn
from torch.autograd import Variable

conv = nn.Conv3d(1,8, kernel_size=3, padding=1, stride = 2)
conv = conv.cuda()
deconv = nn.ConvTranspose3d(8,1, kernel_size=3, stride=2, output_padding=-1)
deconv = deconv.cuda()

x = Variable(torch.randn(8,1,64,64,64).cuda(), requires_grad=True)
#ok
for i in range(10):
    out = conv(x)
    err = out.sum()
    err.backward()

print("deconv")

#error
x = Variable(torch.randn(8,8,32,32,32).cuda(), requires_grad=True)

for i in range(10):
    out = deconv(x)
    err = out.sum()
    err.backward()
```
The issue is, in the deconvolution case padding is not passed to convolution descriptor, ltraces for the failing and passing cudnnGetConvolutionBackwardDataAlgorithm below, padding is 1 for passing and 0 for failing. 
```
[pid 21910] _C.cpython-35m-x86_64-linux-gnu.so->cudnnGetConvolutionBackwardDataAlgorithm(0xffff7fffffff, { CUDNN_DATA_FLOAT, 0, [ 5, 8, 1, 3... ], CUDNN_TENSOR_NCHW }, { CUDNN_DATA_FLOAT, 0, 5, [ 2097152, 0, 8, 8... ], [ 0, 0, 262144, 32768... ] }, { CUDNN_CROSS_CORRELATION, CUDNN_DATA_FLOAT, 3, [ 0, 0, 0, 0... ], [ 2, 2, 2, 0... ], [ 1, 1, 1, 0... ] }, { CUDNN_DATA_FLOAT, 0, 5, [ 2097152, 0, 8, 1... ], [ 0, 0, 262144, 262144... ] }, 1, 0, 0 <unfinished ...>
[pid 21917] _C.cpython-35m-x86_64-linux-gnu.so->cudnnGetConvolutionBackwardDataAlgorithm(0xffff7fffffff, { CUDNN_DATA_FLOAT, 0, [ 5, 8, 1, 3... ], CUDNN_TENSOR_NCHW }, { CUDNN_DATA_FLOAT, 0, 5, [ 2097152, 0, 8, 8... ], [ 0, 0, 262144, 32768... ] }, { CUDNN_CROSS_CORRELATION, CUDNN_DATA_FLOAT, 3, [ 1, 1, 1, 0... ], [ 2, 2, 2, 0... ], [ 1, 1, 1, 0... ] }, { CUDNN_DATA_FLOAT, 0, 5, [ 2097152, 0, 8, 1... ], [ 0, 0, 262144, 262144... ] }, 1, 0, 0 <unfinished ...>
```
Also, there seems to be no way of disabling cudnn at runtime, torch.backends.cudnn.enabled is no longer checked for convolutions: https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/functions/convolution.cpp#L108-L110

**Comment by ngimel** [comment]
It turns out that cudnn breaks only for negative values of output_padding. While I certainly could add checks for that, does that mean that regular padding should be checked for negative values too? Cudnn would break in this case also (regular padding being negative), and no checks are performed currently. @albanD, @abweiss, how common is negative padding in transposed convolutions?

**Comment by soumith** [comment]
yes, could you add negative checks for regular padding too, thank you.
In transposed convolutions, negative padding is common (i wouldn't say fairly common)
